File: string_permutation_ii/jz_recursive.py
#adapted jz recursive java, , use dfs recursive template
#Your submission beats 87.11% Submissions!
class Solution:
    def stringPermutation2(self, str1):
        result = []
        if str1 == None:
            return result
        if len(str1) == 0:
            result.append("")
            return result
        #clist = [c for c in str1]
        #clist.sort() #list.sort() return None, but list is sorted 
        #str1 = ''.join( clist )
        #one liner with sorted function, can not use list.sort()
        str1 = ''.join( sorted([c for c in str1]))
        visited = [False for i in range(len(str1))]
        self.dfs(str1, "", result, visited)
        return result

    def dfs(self, str1, subset, result, visited):
        if len(subset) == len(str1):
            #exit of recursion
            result.append(subset)
            return

        for i  in range( len(str1) ):
            if visited[i]:
                continue

            if (i > 0 and str1[i]== str1[i - 1] and not visited[i - 1]):
                #remove duplicates
                continue
    
            visited[i] = True
            self.dfs(str1, subset + str1[i], result, visited)
            visited[i] = False

File: string-permutation-ii/jz_recursive.java
//jz java, recursive, use dfs recursive template
//Your submission beats 74.20% Submissions!

public class Solution {
    public List<String> stringPermutation2(String str) {
        List<String> result = new ArrayList<>();
        if (str == null) {
            return result;
        } 
        if (str.length() == 0) {
            result.add("");
            return result;
        }
        
        //sort string
        str = sortString(str);
        
        dfs(str, "", result, new boolean[str.length()]);
        return result;
    }
    
    private void dfs(String str, String subset, List<String> result, boolean[] visited) {
        if (subset.length() == str.length()) {
            //exit of recursion
            result.add(subset);
            return;
        }
        
        for (int i = 0; i < str.length(); i++) {
            if (visited[i]) {
                continue;
            }
            
            if (i > 0 && str.charAt(i) == str.charAt(i - 1) && !visited[i - 1]) {
                //remove duplicates
                continue;
            }
            
            visited[i] = true;
            dfs(str, subset + str.charAt(i), result, visited);
            visited[i] = false;
        }
    }
    
    private String sortString(String str) {
        char[] charArray = str.toCharArray();
        Arrays.sort(charArray);
        str = new String(charArray);
        return str;
    }
}
